ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/delay_stubs.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2024 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+#include <caml/memory.h>
+#include <caml/threads.h>
+#include <caml/custom.h>
+#include <caml/unixsupport.h>
+
+#include <stdbool.h>
+#include <errno.h>
+#include <pthread.h>
+
+typedef struct delay {
+	pthread_mutex_t mtx;
+	pthread_cond_t cond;
+	bool signaled;
+} delay;
+
+// Initialize delay
+// Returns error number or 0 if success
+static int delay_init(delay *d)
+{
+	int err;
+	pthread_condattr_t cond_attr;
+
+	d->signaled = false;
+
+	err = pthread_condattr_init(&cond_attr);
+	if (err)
+		goto err0;
+	err = pthread_condattr_setclock(&cond_attr, CLOCK_MONOTONIC);
+	if (!err)
+		err = pthread_cond_init(&d->cond, &cond_attr);
+	if (err)
+		goto err1;
+	err = pthread_mutex_init(&d->mtx, NULL);
```

Comment:
OCaml would use an error checking mutex, might be useful to do the same here to detect bugs.

Reply:
Which bugs would be addressing? Memory corruption? Than we would have much more issues

Reply:
You are comparing a bit apples and pears. The Ocaml runtime set the error check on mutexes exposed to external usage, but not for internal ones (where `lock`/`unlock` are not checked too).
The reason is that the external interface is prone to errors you want to catch.

Reply:
POSIX says:
> If the mutex type is PTHREAD_MUTEX_ERRORCHECK, then error checking shall be provided. If a thread attempts to relock a mutex that it has already locked, an error shall be returned. If a thread attempts to unlock a mutex that it has not locked or a mutex which is unlocked, an error shall be returned.

So it is not memory corruption, but incorrect API use. As long as lock/unlock is always called from the same function that is unlikely to happen in this case, unless unlock is split into another function (in which case it might get accidentally called on the wrong path).